Updated service implementations with transaction support & test files to match transaction implementation to prevent partial QA imports #6559

spetros-do opened the pull request:
### Changes Made:
**Enhanced Transaction Support:** 
- Added transaction parameter support to all service methods that interact with the database

**Improved Error Handling:** 
- Ensured error propagation throughout the transaction chain

**Repository Access :**
- Streamlined how repositories are accessed when a transaction is provided

**Tests:** 
- Updated all test files to match the actual implementation

**Historical:**
- Record Handling: Improved handling of historical records during import operations

spetros-do committed:
Update service implementations with transaction support & test files to match transaction implementation

maxdiebold-erg reviewed (commented):
It looks like your autoformatter changed many of these files from indenting 2 spaces to 4 spaces, so it is very hard to read. Could you configure your formatter for 2 spaces and reformat the affected files?

spetros-do commented:
> It looks like your autoformatter changed many of these files from indenting 2 spaces to 4 spaces, so it is very hard to read. Could you configure your formatter for 2 spaces and reformat the affected files?

Of course. I will advise once done.

spetros-do committed:
(configured the formatter for 2 spaces as per MaxD request)
- Update service implementations with transaction support & test files to match transaction implementation

maxdiebold-erg commented on src/utilities/constants.ts:
```diff
@@ -1,4 +1,45 @@
+  });
+
+  // If there were errors, log a summary
+  if (errors.length > 0 && logger) {
+    logger.error(`${errors.length} errors occurred while processing promises`);
+  }
```
This should throw the errors that are found. As it is now, errors caught here do not cause the transaction to abort, and a success message is incorrectly returned to the client:
```suggestion
  if (errors.length > 0 && logger) {
    logger.error(`${errors.length} errors occurred while processing promises`);
    throw errors[0];
  }
```

maxdiebold-erg commented on src/qa-certification-workspace/qa-certification.service.ts:
```diff
@@ -35,152 +51,296 @@ export class QACertificationWorkspaceService {
+        }
+
+        // If rethrowErrors is true and there are errors, throw the first one
+        if (rethrowErrors && errors.length > 0) {
+          throw errors[0];
+        }
```
This `rethrowErrors` flag should not be here: the errors should always be re-thrown so that that the transaction is aborted.
```suggestion
        if (errors.length > 0) {
          throw errors[0];
        }
```

maxdiebold-erg commented on src/test-summary-workspace/test-summary.service.ts:
```diff
@@ -702,8 +744,14 @@ export class TestSummaryWorkspaceService {
-    await this.repository.save(entity);
-    const result = await this.repository.getTestSummaryById(entity.id);
+    // Use the transaction entity manager if provided
+    const repo = trx ? trx.getRepository(this.repository.target) : this.repository;
+    await repo.save(entity);
+
+    // Use the transaction entity manager if provided
+    const result = await (trx
+      ? trx.getRepository(this.repository.target).findOne({where: {id: entity.id}})
+      : this.repository.getTestSummaryById(entity.id));
```
The `findOne` call is not using the required relations, resulting a null access error in `this.map.one` below. This can be fixed by using the same method to retrieve the new entity (`getTestSummaryById`).

However, this is not possible using the current method of associating a transaction with a repository, since it returns `Repository<TestSummary>` while `getTestSummaryById` is on the child class `TestSummaryWorkspaceRepository`. There may be a workaround for this, but the approach I recommended in the ticket is to use the custom `withTransaction` function to retrieve the transactional repository.

```suggestion
    const repo = withTransaction(this.repository, trx);
    await repo.save(entity);

    // Use the transaction entity manager if provided
    const result = await repo.getTestSummaryById(entity.id);
```

Where `withTransaction` is defined as follows:
```typescript
export function withTransaction<E, T extends Repository<E>>(
  repository: T,
  trx?: EntityManager,
) {
  if (!trx) return repository;

  const repositoryConstructor = repository.constructor as {
    new (manager: EntityManager): T;
  };

  const { target, manager, queryRunner, ...otherRepositoryProperties } =
    repository;

  return Object.assign(new repositoryConstructor(trx), {
    ...otherRepositoryProperties,
  });
}
```

maxdiebold-erg reviewed (changes requested):
I'm still looking through this, but I've found a few bugs in initial testing, which I've noted below.

So far, it looks like the transaction is being used correctly in queries that modify data. There are many cases where the transaction is not being used, which _should_ be fine if the data is only being read and if the results are unaffected by other queries in the transaction. On the flip side, the desire was expressed during performance testing that the number of connections used during imports be limited (i.e. by sending **all** queries through the transaction, not just mutating ones), but connection settings have been tweaked since then so this may be less of an issue.

maxdiebold-erg commented on src/app-e-heat-input-from-gas-workspace/app-e-heat-input-from-gas-workspace.service.ts:
```diff
@@ -81,26 +84,45 @@ export class AppEHeatInputFromGasWorkspaceService {
-    await this.repository.save(entity);
-
-    entity = await this.repository.getAppEHeatInputFromGasById(entity.id);
+    if (trx) {
+      entity = trx.getRepository(AppEHeatInputFromGas).create({
+        id: historicalRecordId ? historicalRecordId : uuid(),
+        appECorrTestRunId,
+        monitoringSystemId: system.id,
+        gasVolume: payload.gasVolume,
+        gasGCV: payload.gasGCV,
+        gasHeatInput: payload.gasHeatInput,
+        userId,
+        addDate: timestamp,
+        updateDate: timestamp,
+      });
+
+      await trx.getRepository(AppEHeatInputFromGas).save(entity);
+      entity = await trx.getRepository(AppEHeatInputFromGas).findOneBy({id: entity.id});
+    } else {
+      entity = this.repository.create({
+        id: historicalRecordId ? historicalRecordId : uuid(),
+        appECorrTestRunId,
+        monitoringSystemId: system.id,
+        gasVolume: payload.gasVolume,
+        gasGCV: payload.gasGCV,
+        gasHeatInput: payload.gasHeatInput,
+        userId,
+        addDate: timestamp,
+        updateDate: timestamp,
+      });
+
+      await this.repository.save(entity);
+      entity = await this.repository.getAppEHeatInputFromGasById(entity.id);
+    }
```
This should use the `withTransaction` method so that the `getAppEHeatInputFromGasById` method can be used when in a transaction:
```suggestion
    const repository = withTransaction(this.repository, trx);
    entity = repository.create({
      id: historicalRecordId ? historicalRecordId : uuid(),
      appECorrTestRunId,
      monitoringSystemId: system.id,
      gasVolume: payload.gasVolume,
      gasGCV: payload.gasGCV,
      gasHeatInput: payload.gasHeatInput,
      userId,
      addDate: timestamp,
      updateDate: timestamp,
    });

    await repository.save(entity);
    entity = await repository.getAppEHeatInputFromGasById(entity.id);
```

maxdiebold-erg commented on src/app-e-heat-input-from-oil-workspace/app-e-heat-input-from-oil.service.ts:
```diff
@@ -81,29 +84,55 @@ export class AppEHeatInputFromOilWorkspaceService {
+      entity = trx.getRepository(AppEHeatInputFromOil).create({
+        id: historicalRecordId ? historicalRecordId : uuid(),
+        monitoringSystemId: system.id,
+        oilMass: payload.oilMass,
+        oilGCV: payload.oilGCV,
+        oilGCVUnitsOfMeasureCode: payload.oilGCVUnitsOfMeasureCode,
+        oilHeatInput: payload.oilHeatInput,
+        oilVolume: payload.oilVolume,
+        oilVolumeUnitsOfMeasureCode: payload.oilVolumeUnitsOfMeasureCode,
+        oilDensity: payload.oilDensity,
+        oilDensityUnitsOfMeasureCode: payload.oilDensityUnitsOfMeasureCode,
+        appECorrTestRunId,
+        userId,
+        addDate: timestamp,
+        updateDate: timestamp,
+      });
+
+      await trx.getRepository(AppEHeatInputFromOil).save(entity);
+      entity = await trx.getRepository(AppEHeatInputFromOil).findOneBy({id: entity.id});
+    } else {
+      entity = this.repository.create({
+        id: historicalRecordId ? historicalRecordId : uuid(),
+        monitoringSystemId: system.id,
+        oilMass: payload.oilMass,
+        oilGCV: payload.oilGCV,
+        oilGCVUnitsOfMeasureCode: payload.oilGCVUnitsOfMeasureCode,
+        oilHeatInput: payload.oilHeatInput,
+        oilVolume: payload.oilVolume,
+        oilVolumeUnitsOfMeasureCode: payload.oilVolumeUnitsOfMeasureCode,
+        oilDensity: payload.oilDensity,
+        oilDensityUnitsOfMeasureCode: payload.oilDensityUnitsOfMeasureCode,
+        appECorrTestRunId,
+        userId,
+        addDate: timestamp,
+        updateDate: timestamp,
+      });
+
+      await this.repository.save(entity);
+      entity = await this.repository.getAppEHeatInputFromOilById(entity.id);
+    }
```
This should use the `withTransaction` method (described in another comment) so that the `getAppEHeatInputFromOilById` method can be used even when in a transaction:
```suggestion
    const repository = withTransaction(this.repository, trx);
    entity = repository.create({
      id: historicalRecordId ? historicalRecordId : uuid(),
      monitoringSystemId: system.id,
      oilMass: payload.oilMass,
      oilGCV: payload.oilGCV,
      oilGCVUnitsOfMeasureCode: payload.oilGCVUnitsOfMeasureCode,
      oilHeatInput: payload.oilHeatInput,
      oilVolume: payload.oilVolume,
      oilVolumeUnitsOfMeasureCode: payload.oilVolumeUnitsOfMeasureCode,
      oilDensity: payload.oilDensity,
      oilDensityUnitsOfMeasureCode: payload.oilDensityUnitsOfMeasureCode,
      appECorrTestRunId,
      userId,
      addDate: timestamp,
      updateDate: timestamp,
    });

    await repository.save(entity);
    entity = await repository.getAppEHeatInputFromOilById(entity.id);
```

maxdiebold-erg commented on src/fuel-flow-to-load-test-workspace/fuel-flow-to-load-test-workspace.service.ts:
```diff
@@ -108,7 +116,7 @@ export class FuelFlowToLoadTestWorkspaceService {
     entity.testBasisCode = payload.testBasisCode;
     entity.averageDifference = payload.averageDifference;
-    entity.numberOfHoursUsed = payload.numberOfHoursExcludedCofiring;
+    entity.numberOfHoursUsed = payload.numberOfHoursUsed;
```
I agree that this was probably a mistake and should be corrected, but I think we may want to include `numberOfHoursExcludedCofiring`, too:
```suggestion
    entity.numberOfHoursUsed = payload.numberOfHoursUsed;
    entity.numberOfHoursExcludedCofiring = payload.numberOfHoursExcludedCofiring;
```

maxdiebold-erg commented on src/rata-summary-workspace/rata-summary-workspace.service.ts:
```diff
@@ -72,7 +75,10 @@ export class RataSummaryWorkspaceService {
     // Checks if RATA Test is valid.
     await this.rataService.getRataById(rataId);
```
I'm unable to select the above lines for commenting, but they will need to be updated, too. Both methods in the preceding four lines should receive the transaction (`trx`):
```diff
    // Checks if Test Summary is valid.
-    await this.testSummaryService.getTestSummaryById(testSumId);
+    await this.testSummaryService.getTestSummaryById(testSumId, trx);
    // Checks if RATA Test is valid.
-    await this.rataService.getRataById(rataId);
+    await this.rataService.getRataById(rataId, trx);
```
I've made comments on each of those methods about using the transaction.

maxdiebold-erg commented on src/rata-workspace/rata-workspace.service.ts:
```diff
@@ -29,7 +30,8 @@ export class RataWorkspaceService {
+  ) {
+  }
 
   async getRataById(id: string): Promise<RataDTO> {
     const result = await this.repository.findOneBy({
```
This should be updated to accept a transaction and conditionally use it:
```suggestion
  async getRataById(id: string, trx?: EntityManager): Promise<RataDTO> {
    const result = await (trx
      ? trx.getRepository(this.repository.target)
      : this.repository
    ).findOneBy({
```

maxdiebold-erg commented on src/test-summary-workspace/test-summary.service.ts:
```diff
@@ -91,7 +111,8 @@ export class TestSummaryWorkspaceService {
+  ) {
+  }
 
   async getTestSummaryById(testSumId: string): Promise<TestSummaryDTO> {
     const result = await this.repository.getTestSummaryById(testSumId);
```
This should be updated to accept a transaction and conditionally use it (it should use the `withTransaction` utility since it calls a custom repository method):
```suggestion
  async getTestSummaryById(testSumId: string, trx?: EntityManager): Promise<TestSummaryDTO> {
    const result = await withTransaction(this.repository, trx).getTestSummaryById(testSumId);
```

maxdiebold-erg commented on src/test-qualification-workspace/test-qualification-workspace.service.ts:
```diff
@@ -128,6 +139,8 @@ export class TestQualificationWorkspaceService {
     }
 
     record.testClaimCode = payload.testClaimCode;
+    // Set the TestClaimCode relation to null to avoid TypeScript error
+    record.TestClaimCode = null;
```
I don't see any TypeScript error when removing this, and adding it may cause a DB error.
```suggestion
```

maxdiebold-erg commented on src/test-qualification-workspace/test-qualification-workspace.service.ts:
```diff
@@ -165,9 +179,10 @@ export class TestQualificationWorkspaceService {
     payload: TestQualificationImportDTO,
     userId: string,
     isHistoricalRecord: boolean,
+    trx?: EntityManager,
   ) {
     const isImport = true;
-    let historicalRecord: TestQualification;
+    let historicalRecord: any; // Use any type to avoid TypeScript errors
```
The type error is due to `TestQualification` referring to the "workspace" version of the entity, when the "official" is what's queried here. This can be fixed by adding
```typescript
import {TestQualification as TestQualificationOfficial} from '../entities/test-qualification.entity';
```
to the imports, then updating this line to
```typescript
    let historicalRecord: TestQualificationOfficial;
```

maxdiebold-erg commented on src/unit-default-test-run-workspace/unit-default-test-run.service.ts:
```diff
@@ -66,10 +67,13 @@ export class UnitDefaultTestRunWorkspaceService {
     const timestamp = currentDateTime();
 
-    let entity = this.repository.create({
+    const repo = trx ? trx.getRepository(UnitDefaultTestRun) : this.repository;
```
In calling `trx.getRepository(UnitDefaultTestRun)`, **you are targeting the official repository when creating a new entity** (since `this.repository` points to the "workspace" version).

This is another reason why I recommend using the `withTransaction` utility, instead. I only caught this at the end of this review, so this may be an issue elsewhere, too. I would recommend going back through all other locations that use this pattern and either: A) switch to the custom `withTransaction` utility or B) ensure the repositories align.

maxdiebold-erg reviewed (changes requested):
I finished reviewing the code and added some comments, but it wasn't until reviewing the changes in **`UnitDefaultTestRunWorkspaceService`** that I caught a major and prevalent issue: in employing the following common pattern:
```typescript
const repo = trx ? trx.getRepository(UnitDefaultTestRun) : this.repository;
```
In this example and several others, the entity representing the table in "official" is being used when a transaction is passed, while the "workspace" table should be used.

In other cases, this is correctly coded as:
```typescript
const repo = trx ? trx.getRepository(this.repository.target) : this.repository;

// Alternatively:
const repo = withTransaction(this.repository, trx);
```

I haven't gone back through the changes again yet, but I suggest updating all uses of such pattern to use either `this.repository.target` or `withTransaction` rather than a bare entity class.

mark-hayward-erg commented:
closing this old/obsolete PR
